Sources/ConcurrencyTaskManager/TaskStackActor.swift

```diff
@@ -0,0 +1,155 @@
+import DequeModule
+import Foundation
+
+public actor TaskStackActor {
+
+  private struct State {
+    var waitingCount: Int = 0
+    var executingCount: Int = 0
+
+    mutating func update(waitingCount: Int, executingCount: Int) {
+      self.waitingCount = waitingCount
+      self.executingCount = executingCount
+    }
+  }
+
+  private var stack: Deque<TaskNode> = .init()
+  private var executings: ContiguousArray<TaskNode> = .init()
+
+  public private(set) var maxConcurrentTaskCount: Int
+
+  private var currentExecutingTaskCount: Int = 0
+
+  @Published private var state: State = .init()
+
+  public init(maxConcurrentTaskCount: Int) {
+    self.maxConcurrentTaskCount = maxConcurrentTaskCount
+  }
+
+  deinit {
+
+  }
+
+  @discardableResult
+  public func addTask<Return>(
+    label: String = "",
+    priority: TaskPriority? = nil,
+    operation: @escaping @Sendable () async throws -> Return
+  ) -> Task<Return, Error> {
+
+    let extendedContinuation: AutoReleaseContinuationBox<Return> = .init(nil)
+
+    let referenceTask = Task { [weak extendedContinuation] in
+      return try await withUnsafeThrowingContinuation{ (continuation: UnsafeContinuation<Return, Error>) in
+        extendedContinuation?.setContinuation(continuation)
+      }
+    }
+
+    let newNode = TaskNode(label: label) { [weak self] box in
+
+      // Run the operation
+
+      await withTaskCancellationHandler {
+        do {
+          let result = try await operation()
+
+          guard Task.isCancelled == false else {
+            extendedContinuation.resume(throwing: CancellationError())
+            return
+          }
+
+          extendedContinuation.resume(returning: result)
+
+        } catch {
+
+          guard Task.isCancelled == false else {
+            extendedContinuation.resume(throwing: CancellationError())
+            return
+          }
+
+          extendedContinuation.resume(throwing: error)
+
+        }
+      } onCancel: {
+        referenceTask.cancel()
+      }
+
+      await self?.decrement()
+
+      // connecting to the next if presents
+
+      if let node = box.value {
+        await self?.processForCompletion(taskNode: node)
+      }
+
+      await self?.drain()
+
+    }
+
+    stack.prepend(newNode)
+
+    drain()
+
+    return referenceTask
+  }
+
+  /**
+   Waits until the all enqueued are processed.
+   Including added items while processing.
+   */
+  public func waitUntilAllItemProcessed() async {
+
+    let stream = AsyncStream<State> { continuation in
+
+      let cancellable = $state.sink { state in
+        continuation.yield(state)
+      }
+
+      continuation.onTermination = { _ in
+        cancellable.cancel()
+      }
+
+    }
+
+    for await state in stream {
+      if state.executingCount == 0 && state.waitingCount == 0 {
+        break
+      }
+    }
+
+  }
+
+  private func increment() {
+    currentExecutingTaskCount += 1
```

Comment:
Make the state update inside in the same way as the decrement method

Reply:
yes, but it happens here.
I might as well install this operation there and then remove this function.

https://github.com/VergeGroup/swift-concurrency-task-manager/pull/5/files#diff-4701dbc7cc90ab860cbfa4d49a9d770499ebe2f459e81bbbdcb9996f063e45f1R149